### ISSUE TITLE
Negative number in Points for stars received

### ISSUE THREAD

**Issue by lucaboesch** [frankkoch/moodle-mod_studentquiz#284]
Hi Frank and the StudentQuiz team

One of our students came asking.
How come I have a negative score in Points for stars received?
I thought I share this one with you and maybe you've got an idea.

![grafik](https://user-images.githubusercontent.com/377279/97678214-c1456f80-1a93-11eb-9973-0e4bcd353501.png)

Best,
Luca

**Comment by frankkoch** [comment]
This doesn't make sense, but could it be that the points awarded for stars reveived have been set to negative?

**Comment by lucaboesch** [comment]
Should have added this was the first thing I cheked.
No the values are plain regularly set.
<img width="542" alt="grafik" src="https://user-images.githubusercontent.com/377279/97680598-3c5c5500-1a97-11eb-940b-bc46695ae0a6.png">

**Comment by dionysius** [comment]
That is weird. Even when only receiving 1 stars, it should be positive. I'm now thinking if its possible to construct the url with a negative amount of stars...
Can you check the studenquiz_rate table if there's a rate negative?

**Comment by dionysius** [comment]
Unfortunately it would be possible that someone did sent an invalid amount of stars. The only check is if it's integer and converted to an integer: https://github.com/frankkoch/moodle-mod_studentquiz/blob/master/save.php#L63

**Comment by lucaboesch** [comment]
Unfortunately that activity went on and the value is not any more as to the point in time the user reported it. So my checking the tables now would be useless.

**Comment by dionysius** [comment]
Doesn't need to be. The table studentquiz_rate consists of the single rating of a user to a question. The ranking is the calculated total. So if the student hasn't rerated this exact question, that entry can still be found there. But in any way, I've marked this issue as bug since we should really limit the allowed rating number in the backend.

**Comment by lucaboesch** [comment]
Okay I'll take a look then.

**Comment by lucaboesch** [comment]
Nope, I don't see any.

`SELECT * FROM mdl_studentquiz_rate where rate not in (1,2,3,4,5);`
` id | rate | questionid | userid `
`----+------+------------+--------`
`(0 rows)`

**Comment by machiaveli88** [comment]
Hello, I've the same problem:
<img width="1346" alt="Bildschirmfoto 2020-11-05 um 07 17 09" src="https://user-images.githubusercontent.com/12184856/98204974-774b0680-1f37-11eb-97f4-432d23fb2ba8.png">
<img width="1346" alt="Bildschirmfoto 2020-11-05 um 07 17 32" src="https://user-images.githubusercontent.com/12184856/98204978-79ad6080-1f37-11eb-9d25-a7887b6ac9cf.png">
<img width="1346" alt="Bildschirmfoto 2020-11-05 um 07 18 18" src="https://user-images.githubusercontent.com/12184856/98204981-7ade8d80-1f37-11eb-8779-921793865c75.png">

Some of our students are getting a negative amount of points for their ratings. I've checked this with the red marked entry/student. He submitted four questions, one of them marked as confirmed and one marked as rejected.
The negative amount of points is exactly the twice than the the community rating (I've checked this by changing my rating), so I think it's a bug in the calculation.

**Comment by machiaveli88** [comment]
Okay, I think I've got it. Another student has 5 questions: 1 confirmed, 1 rejected, 3 open. The one confirmed question has a rating of 3.95 and he got -11.85 points for rating (so 3x). if I confirm another one (so 2 confirmed, 1 rejected, 2 open), the points are changing to -7.9 (so 2x). So I think it depends on the open (not confirmed/rejected) questions.

**Comment by lucaboesch** [comment]
I would be looking forward for a solution not punishing students for sumitting a question which then lay in an 'open' status in any way.
Because from a pedagogical point of view that would be a massive disencouragement of submitting a question to StudentQuiz.

**Comment by dionysius** [comment]
There is definitely a calculation off. It looks like a rating on a rejected question has a multiplicative negative impact on the score. That's definitely not intended

**Comment by dionysius** [comment]
I've difficulties to get it reproduced. The lowest level part of this calculation data comes from https://github.com/frankkoch/moodle-mod_studentquiz/blob/master/locallib.php#L925. This query is basically a short version of it:
```sql
SELECT
	q.id,
	q.createdby AS createdby,
	AVG(sqv.rate) AS avg_rate_perq,
	COUNT(sqv.rate) AS num_rate_perq,
	MAX(CASE WHEN sqv.id IS NULL THEN 1 ELSE 0 END) AS question_not_rated
FROM
	mdl_question q
LEFT JOIN mdl_studentquiz_rate sqv ON
	q.id = sqv.questionid
WHERE
	q.hidden = 0
	AND q.parent = 0
        AND q.createdby = ...
GROUP BY
	q.id,
	q.createdby
```
Can you find those values for a problematic user? I'd like to see how those values do and what is done with them.

**Comment by Peyje** [comment]
I'm pretty sure I found the issue and corresponding fix.

Taking a look at [locallib.php#L904](https://github.com/frankkoch/moodle-mod_studentquiz/blob/master/locallib.php#L904), questions are **not** counted into the *countq* counter (to my understanding the counter for the amount of questions a student created) if the question itself is hidden (which is the case as long as a question is not yet approved/rejected).

So suppose a student has written three questions and only one of those is already approved with an average rating of 4 points. Lets take a look at this line [locallib.php#L834](https://github.com/frankkoch/moodle-mod_studentquiz/blob/master/locallib.php#L834):
`COALESCE(rates.avgv, 0) * (COALESCE(creators.countq, 0) -
                                                   COALESCE(rates.not_rated_questions, 0)) * :ratequantifier`
This then resolves to
`4 * (1 - 2) * :ratequantifier`
as the hidden questions are not counted into *countq*.

My proposed solution is to just remove the line [llocallib.php#L904](https://github.com/frankkoch/moodle-mod_studentquiz/blob/master/locallib.php#L904) (and while we are at it, line [locallib.php#L921](https://github.com/frankkoch/moodle-mod_studentquiz/blob/master/locallib.php#L921) as well, as approved questions also only count if not hidden right now).

Please be sure to verify this before approving it, as my knowledge about this plugin is not too extensive.
If this can be reproduced, I already created a fix, see Pull Request #291.

**Comment by zabellemotte** [comment]
You analyze, Peyje, matches with what we observe in our Moodle : students that have nagative points are those who submitted questions that are not yet validated.

**Comment by dionysius** [comment]
Thank you all for your input! I'm sure the keyword hidden was what I was missing trying to reproduce the issue! The default settings in StudentQuiz is publishing all questions.

I still need to think this through completely, but I have a first response for this issue:

The interesting part is `q.hidden` **does not mean** the question got hidden in studentquiz! This q.hidden equals that the question has been deleted in moodle. Moodle doesn't delete questions completely when they're deleted. If that question has been used once somewhere, it just sets the hidden to 1.

Let me elaborate a modified version of [this subquery](https://github.com/frankkoch/moodle-mod_studentquiz/blob/master/locallib.php#L930):

```sql
SELECT
	q.id,
	q.createdby AS createdby,
	AVG(sqv.rate) AS avg_rate_perq,
	COUNT(sqv.rate) AS num_rate_perq,
	MAX(CASE WHEN sqv.id IS NULL THEN 1 ELSE 0 END) AS question_not_rated,
	q.hidden as core_hidden,
	sqq.hidden as sqq_hidden
FROM mdl_studentquiz sq
JOIN mdl_context con ON con.instanceid = sq.coursemodule
JOIN mdl_question_categories qc ON qc.contextid = con.id
JOIN mdl_question q ON q.category = qc.id
LEFT JOIN mdl_studentquiz_rate sqv ON q.id = sqv.questionid
-- new join
JOIN mdl_studentquiz_question sqq ON q.id = sqq.questionid 
-- and temporarely remove that hidden filter
-- WHERE q.hidden = 0
--    AND q.parent = 0
WHERE q.parent = 0
	AND sq.coursemodule = 16
GROUP BY
	q.id,
	q.createdby
```

My result:
|id|createdby|avg_rate_perq|num_rate_perq|question_not_rated|core_hidden|sqq_hidden|
|--|---------|-------------|-------------|------------------|-----------|----------|
|32|2|1|1|0|0|1|
|33|2|1|1|0|0|1|
|34|2|1|1|0|0|1|
|35|2||0|1|1|0|
|36|2|1|1|0|0|1|
|37|2|1|1|0|0|1|
|38|2|1|1|0|0|1|
|39|2|1|1|0|0|1|
|40|2|1|1|0|0|1|
|41|2|5|1|0|0|0|
|42|2|1|1|0|0|1|
|43|2|1|1|0|0|1|
|44|2|1|1|0|1|1|
|45|2||0|1|1|0|

I have 11 visible questions in studentquiz (core_hidden = 0) and 3 questions which actually are deleted in moodle (core_hidden = 1). The visible questions are all hidden in studentquiz (sqq_hidden = 1) except one (sqq_hidden = 0). 

My first impression is, that this subquery is missing a `AND sqq.hidden = 0`

**Comment by dionysius** [comment]
My proposed change is currently:

```diff
diff --git a/locallib.php b/locallib.php
index 3b0a7f4..1f4451e 100644
--- a/locallib.php
+++ b/locallib.php
@@ -934,9 +934,11 @@ function mod_studentquiz_helper_attempt_stat_joins($excluderoles=array()) {
                                JOIN {context} con ON con.instanceid = sq.coursemodule
                                JOIN {question_categories} qc ON qc.contextid = con.id
                                JOIN {question} q ON q.category = qc.id
+                               JOIN {studentquiz_question} sqq ON q.id = sqq.questionid
                           LEFT JOIN {studentquiz_rate} sqv ON q.id = sqv.questionid
                               WHERE q.hidden = 0
                                     AND q.parent = 0
+                                    AND sqq.hidden = 0
                                     AND sq.coursemodule = :cmid6
                            GROUP BY q.id, q.createdby
                            ) avgratingperquestion
```

**Comment by Peyje** [comment]
I can confirm that your proposed change resolves the issue with the negative ratings as well.

On a side note however, questions that are still hidden (so `sqq.hidden = 1`, which happens if it is not yet approved/rejected if the option _"Needs approval before publication"_ is set) will then still not give any points for a created question. While this does not lead to an error in this calculation, it begs the question if this is supposed be handled that way.

Suppose a question is created but then rejected, then the student should get the points for a created question, while not getting points for an approved question, right? With your solution, he or she would get zero points, as the question always remains hidden.

This is in line with how it was handled before as well, but I would argue that my solution in PR #291 (count the questions no matter the state of `sqq.hidden`) would make more sense in this particular situation.

I guess the question then simply is: How should hidden questions be handled?

**Comment by dionysius** [comment]
> I guess the question then simply is: How should hidden questions be handled?

That's exactly how I first started my response while digging deeper into the cause. I also have to dig into my archives to find out what it was initially planned to do. I just know, that the hidden-feature came much later that the first definition of the calculation (probably also in line with the new states of the question?). So we probably just happen to hit a spot where additional features forces us to rethink about the past models. And that does mean that me might want to redefine that model to fit the current situation.

Let me summarize some questions which can be answered yes/no:
- In a StudentQuiz without automatic publishing, should the student get a point for a question that he just created and is hidden?
- Should a student get points for a question that is approved but hidden?
- Should a student get points for a question (because of no. questions created) that is disapproved but visible to others?
- Should there be a new factor for disapproved questions that can be set negative (to make up the previous question maybe)? (Analog the factor for last wrong attempts if set to negative)
- Should ratings on hidden question (because it got hidden later) count towards the ranking?
- Should hidden questions of other students count towards the personal progress (must be no, as he's never able to see the question, but still wanted to point that out)

You are correct, with my fix, the calculation basically treats any hidden question as non-existent.

**Comment by Peyje** [comment]
Those are some very good questions that are not easy to answer. Mileage might vary for different understandings of the hidden-feature.

I do not know how you prefer the handling of this, but might I suggest the following:

- Transferring these questions to another, new issue (maybe with the label _enhancement_)
- Closing this issue with your solution and issuing a release to be able to offer a quick fix to the negative points issue
- Closing my PR without merging as this would already change the handling of hidden questions, which should be further discussed

**Comment by dionysius** [comment]
Sounds like a good plan, I'll set it up

**Comment by dionysius** [comment]
immediate fix prepared in #296. Further discussions over new ticket (this will be edited with the issue link)
